ykrt/src/deopt.rs

```diff
@@ -255,15 +255,25 @@ unsafe extern "C" fn __ykrt_deopt(
         }
     }
 
-    unsafe { framerec.reconstruct_frames(frameaddr) }
+    // FIXME: This print doesn't really serve a purpose anymore, since `deoptimise` will always be
+    // followed by `exit-jit-code` as there's no other way to leave a trace.
+    #[cfg(feature = "yk_jitstate_debug")]
+    print_jit_state("exit-jit-code");
+
+    let (ptr, btmframesize) = unsafe { framerec.reconstruct_frames(frameaddr) };
+    // Calculate the offset on the stack we want to write the new frames to. We want to write them
+    // immediately after the frame containing the control point.
+    let newframesdst = frameaddr.sub(btmframesize);
+    __ykrt_reconstruct_frames(ptr as *const c_void, newframesdst);
+    unreachable!();
 }
 
 /// After a guard failure, reconstructs the stack frames and registers and then jumps back to the
 /// AOT code from where to continue.
 #[cfg(target_arch = "x86_64")]
 #[naked]
 #[no_mangle]
-extern "C" fn __ykrt_reconstruct_frames(newframesptr: *const c_void) {
+extern "C" fn __ykrt_reconstruct_frames(newframesptr: *const c_void, newframedst: *const c_void) {
```

Comment:
What was the outcome of the `-> !` discussion?

Reply:
Not sure, @ltratt alluded in the call there's problems with it, so I wanted to get his opinion first.

Reply:
I think it *should* work here, so just try it. If it compiles, it works!

Reply:
[88c2a64](https://github.com/ykjit/yk/pull/762/commits/88c2a64d7344ad43c881c1969b10c4f49b8a217b)